### DISCUSSION TITLE
Generate syntax highlighting files

### DISCUSSION THREAD

**Discussion by vanillajonathan** [zesterer/chumsky#276] [Ideas]
After you create a parser then you have write syntax highlightning files to syntax highlight the language your parser parses. When doing changes to the parser you manually have to update these files. It would be good of Chumsky could generate such syntax highlight files for:
- [Ace](https://github.com/ajaxorg/ace) JavaScript editor.
- [Lezer](https://lezer.codemirror.net/) for the [CodeMirror](https://codemirror.net/) JavaScript editor.
- [Monaco](https://microsoft.github.io/monaco-editor/), Microsoft's JavaScript editor used by VS Code.
- [GtkSourceView](https://gitlab.gnome.org/GNOME/gtksourceview/), the GTK source code editor component used by GNOME Text Editor, GNOME Builder and others.
- [KDE Syntax Highlighting Engine](https://invent.kde.org/frameworks/syntax-highlighting) used by Kate, KDevelop and others. ([API](https://api.kde.org/frameworks/syntax-highlighting/html/index.html))
- [highlight.js](https://highlightjs.org/), syntax highlighting for the web.
- [Sublime Text](https://www.sublimetext.com/docs/syntax.html), text editor
- [TextMate](https://macromates.com/manual/en/language_grammars), text editor

**Comment by zesterer** [comment]
I have considered implementing a system that can generate something like BNF from combinators automatically.

However, not all concepts present in chumsky translate to BNF (especially context-sensitive stuff like that which was implemented in https://github.com/zesterer/chumsky/pull/269), and this is even more true with simpler regex-based syntax highlighters which are only capable of recognising regular grammars.

Besides: it's very rare that syntax highlighters actually parse the input in a similar way to a compiler. They usually just emulate a basic subset of the language's tokeniser/lexer, allowing them to highlight specific tokens without having a structural understanding of the code.

You might have better luck trying to implement something like this on top of [`logos`](https://github.com/maciejhirsz/logos), which is much closer to most syntax highlighting systems in nature (indeed, I even used it in my own [text editor](https://github.com/zesterer/zte) for this very purpose).

**Reply by vanillajonathan** [reply]
Alright thanks for your reply, you may close this discussion if generating such files is deemed not possible or out-of-scope for this project.

**Reply by zesterer** [reply]
I wouldn't say that it's out of scope, more that it's unlikely to be a big focus for a while (unless someone wants to look into it themselves).

**Reply by robo9k** [reply]
Do you happen to know if there is an "universal" API for highlighting instead of individual grammar syntax configuration files?
Quite a bunch of tooling seems to settle on `tree-sitter` nowadays, but that's not a [standard](https://xkcd.com/927/).
LSP does have ["semantic tokens"](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens) but those seem to be in addition to highlighting done locally by the client.

**Reply by zesterer** [reply]
I've not heard of one, no.
